retrieval/scrape.go

```diff
@@ -0,0 +1,243 @@
+type Scraper interface {
+	// Sends samples to ch until an error occurrs, the context is canceled
+	// or no further samples can be retrieved. The channel is closed before returning.
+	Scrape(ctx context.Context, ch chan<- *model.Sample) error
+}
+
+// Scrape pools runs scrapers for a set of targets where a target, identified by
+// a label set is scraped exactly once.
+type ScrapePool struct {
+	appender storage.SampleAppender
+
+	ctx    context.Context
+	cancel func()
+
+	mtx    sync.RWMutex
+	loops  map[model.Fingerprint]*scrapeLoop
+	states map[model.Fingerprint]*TargetStatus
+}
+
+func NewScrapePool(app storage.SampleAppender) *ScrapePool {
+	sp := &ScrapePool{
+		appender: app,
+		states:   map[model.Fingerprint]*TargetStatus{},
+	}
+	sp.ctx, sp.cancel = context.WithCancel(context.Background())
+
+	return sp
+}
+
+func (sp *ScrapePool) Stop() {
+	if sp.cancel != nil {
+		sp.cancel()
+	}
+}
+
+// Sync the running scrapers with the provided list of targets.
+func (sp *ScrapePool) Sync(targets []*Target) {
+	sp.mtx.Lock()
+	defer sp.mtx.Unlock()
+
+	// Shutdown all running loops after their current scrape
+	// is finished.
+	var wg sync.WaitGroup
+	wg.Add(len(sp.loops))
+
+	for _, l := range sp.loops {
+		go func(l *scrapeLoop) {
+			l.stop()
+			wg.Done()
+		}(l)
+	}
+	defer wg.Wait()
+
+	// Old loops are shutting down, create new from the targets.
+	sp.loops = map[model.Fingerprint]*scrapeLoop{}
+
+	for _, t := range targets {
+		fp := t.fingerprint()
+
+		// Create states for new targets.
+		state, ok := sp.states[fp]
+		if !ok {
+			state = &TargetStatus{}
+			sp.states[fp] = state
+		}
+
+		// Create and start loops the first time we see a target.
+		if loop, ok := sp.loops[fp]; !ok {
+			loop = newScrapeLoop(sp.appender, t, state)
+			go loop.run(sp.ctx, t.interval(), t.timeout())
+		}
+	}
+
+	// Drop states of targets that no longer exist.
+	for fp := range sp.states {
+		if _, ok := sp.loops[fp]; ok {
+			delete(sp.states, fp)
+		}
+	}
+}
+
+// scrapeLoop manages scraping cycles.
+type scrapeLoop struct {
+	scraper  Scraper
+	state    *TargetStatus
+	appender storage.SampleAppender
+
+	cancel      func()
+	stopc, done chan struct{}
+	mtx         sync.RWMutex
+}
+
+func newScrapeLoop(app storage.SampleAppender, target *Target, state *TargetStatus) *scrapeLoop {
+	return &scrapeLoop{
+		scraper:  target,
+		appender: target.wrapAppender(app),
+		state:    state,
+	}
+}
+
+func (sl *scrapeLoop) active() bool {
+	return sl.cancel != nil
+}
+
+func (sl *scrapeLoop) run(ctx context.Context, interval, timeout time.Duration) {
+	if sl.active() {
+		return
+	}
+
+	sl.stopc = make(chan struct{})
+	sl.done = make(chan struct{})
+	defer close(sl.done)
+
+	ctx, sl.cancel = context.WithCancel(ctx)
+
+	select {
+	case <-time.After(time.Duration(float64(interval) * rand.Float64())):
+		// Continue after random scraping offset.
+	case <-ctx.Done():
+		return
+	case <-sl.stopc:
+		return
+	}
+
+	timer := time.NewTicker(interval)
```

Comment:
That was a timer for a short time, some leftovers remained.

On Thu, Sep 10, 2015 at 2:27 PM Julius Volz <notifications@github.com>
wrote:

> In retrieval/scrape.go
> <https://github.com/prometheus/prometheus/pull/1064#discussion_r39153489>:
>
> > +	sl.stopc = make(chan struct{})
> > +	sl.done = make(chan struct{})
> > +	defer close(sl.done)
> > +
> > +	ctx, sl.cancel = context.WithCancel(ctx)
> > +
> > +	select {
> > +	case <-time.After(time.Duration(float64(interval) * rand.Float64())):
> > +		// Continue after random scraping offset.
> > +	case <-ctx.Done():
> > +		return
> > +	case <-sl.stopc:
> > +		return
> > +	}
> > +
> > +	timer := time.NewTicker(interval)
>
> I'd call it ticker to not cause confusion because the stdlib also
> differentiates between Timers and Tickers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/prometheus/prometheus/pull/1064/files#r39153489>.
>